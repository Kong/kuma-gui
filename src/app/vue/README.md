---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

