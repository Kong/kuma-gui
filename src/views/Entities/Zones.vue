<template>
  <div class="zones">
    <page-header noflex>
      <breadcrumbs />
    </page-header>
    <KEmptyState
      v-if="multicluster === false"
      class="global-api-status"
    >
      <template slot="title">
        <KIcon
          class="kong-icon--centered"
          icon="dangerCircle"
          size="64"
        />
        Kuma is running in Standalone mode.
      </template>
      <template slot="message">
        <p>
          To access this page, you must be running in <strong>Multi-Zone</strong> mode.
        </p>
      </template>
      <template slot="cta">
        <KButton
          to="https://kuma.io/docs/0.6.0/documentation/deployments/"
          target="_blank"
          appearance="primary"
        >
          Learn More
        </KButton>
      </template>
    </KEmptyState>

    <!-- Remote CPs information for when Multicluster is enabled -->
    <FrameSkeleton v-else>
      <DataOverview
        :page-size="pageSize"
        :has-error="hasError"
        :is-loading="isLoading"
        :empty-state="empty_state"
        :display-data-table="true"
        :table-data="tableData"
        :table-data-is-empty="tableDataIsEmpty"
        table-data-function-text="View"
        table-data-row="name"
        @tableAction="tableAction"
        @reloadData="loadData"
      >
        <template slot="pagination">
          <Pagination
            :has-previous="previous.length > 0"
            :has-next="hasNext"
            @next="goToNextPage"
            @previous="goToPreviousPage"
          />
        </template>
      </DataOverview>
      <Tabs
        v-if="isEmpty === false"
        :has-error="hasError"
        :is-loading="isLoading"
        :tabs="tabs"
        initial-tab-override="overview"
      >
        <template slot="tabHeader">
          <div>
            <h3>{{ tabGroupTitle }}</h3>
          </div>
        </template>
        <template slot="overview">
          <LabelList
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div>
              <ul>
                <li
                  v-for="(value, key) in entity"
                  :key="key"
                >
                  <h4>
                    {{ key }}
                  </h4>
                  <p v-if="key === 'status'">
                    <KBadge :appearance="value === 'Offline' ? 'danger' : 'success'">
                      {{ value }}
                    </KBadge>
                  </p>
                  <p v-else>
                    {{ value }}
                  </p>
                </li>
              </ul>
            </div>
          </LabelList>
        </template>
        <template slot="insights">
          <LoaderCard
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
          >
            <div v-if="rawEntity">
              <div
                v-for="(value, key) in rawEntity.zoneInsight.subscriptions"
                :key="key"
                class="overview-stack"
              >
                <h4 class="overview-title">
                  ID: <span class="mono">{{ value.id }}</span>
                </h4>

                <div v-if="value.globalInstanceId || value.connectTime || value.disconnectTime">
                  <h5 class="overview-tertiary-title">
                    General Information:
                  </h5>
                  <ul>
                    <li v-if="value.globalInstanceId">
                      <strong>Global Instance ID:</strong>&nbsp;
                      <span class="mono">{{ value.globalInstanceId }}</span>
                    </li>
                    <li v-if="value.connectTime">
                      <strong>Last Connected:</strong>&nbsp;
                      {{ value.connectTime | readableDate }}
                    </li>
                    <li v-if="value.disconnectTime">
                      <strong>Last Disconnected:</strong>&nbsp;
                      {{ value.disconnectTime | readableDate }}
                    </li>
                  </ul>
                </div>

                <div v-if="value.status">
                  <ul
                    v-if="value.status.stat"
                    class="overview-stat-grid"
                  >
                    <li
                      v-for="(item, label) in value.status.stat"
                      :key="label"
                    >
                      <h6 class="overview-tertiary-title">
                        {{ label | humanReadable }}:
                      </h6>
                      <ul>
                        <li
                          v-for="(k, v) in item"
                          :key="v"
                        >
                          <strong>{{ v | humanReadable }}:</strong>&nbsp;
                          <span class="mono">{{ k | formatValue | formatError }}</span>
                        </li>
                      </ul>
                    </li>
                  </ul>
                </div>
                <KAlert
                  v-else
                  appearance="info"
                  class="mt-4"
                >
                  <template slot="alertIcon">
                    <KIcon icon="portal" />
                  </template>
                  <template slot="alertMessage">
                    There are no Policy statistics for <strong>{{ value.id }}</strong>
                  </template>
                </KAlert>
              </div>
            </div>
          </LoaderCard>
        </template>
        <template slot="yaml">
          <YamlView
            :title="entityOverviewTitle"
            :has-error="entityHasError"
            :is-loading="entityIsLoading"
            :is-empty="entityIsEmpty"
            :content="yamlEntity"
          />
        </template>
      </Tabs>
    </FrameSkeleton>
  </div>
</template>

<script>
import { mapState, mapGetters } from 'vuex'
import { humanReadableDate, getOffset, getSome, stripTimes, camelCaseToWords } from '@/helpers'
import sortEntities from '@/mixins/EntitySorter'
import FrameSkeleton from '@/components/Skeletons/FrameSkeleton'
import PageHeader from '@/components/Utils/PageHeader.vue'
import Breadcrumbs from '@/components/Breadcrumbs.vue'
import Pagination from '@/components/Pagination'
import DataOverview from '@/components/Skeletons/DataOverview'
import Tabs from '@/components/Utils/Tabs'
import YamlView from '@/components/Skeletons/YamlView'
import LabelList from '@/components/Utils/LabelList'
import LoaderCard from '@/components/Utils/LoaderCard'

export default {
  name: 'Zones',
  metaInfo: {
    title: 'Zones'
  },
  components: {
    FrameSkeleton,
    PageHeader,
    Breadcrumbs,
    Pagination,
    DataOverview,
    Tabs,
    YamlView,
    LabelList,
    LoaderCard
  },
  filters: {
    formatValue (value) {
      return value ? parseInt(value).toLocaleString('en').toString() : 0
    },
    readableDate (value) {
      return humanReadableDate(value)
    },
    humanReadable (value) {
      return camelCaseToWords(value)
    },
    formatError (value) {
      if (value === '--') {
        return 'error calculating'
      }

      return value
    }
  },
  mixins: [
    sortEntities
  ],
  data () {
    return {
      isLoading: true,
      isEmpty: false,
      hasError: false,
      entityIsLoading: true,
      entityIsEmpty: false,
      entityHasError: false,
      tableDataIsEmpty: false,
      empty_state: {
        title: 'No Data',
        message: 'There are no Zones present.'
      },
      tableData: {
        headers: [
          { key: 'actions', hideLabel: true },
          { label: 'Status', key: 'status' },
          { label: 'Name', key: 'name' }
        ],
        data: []
      },
      tabs: [
        {
          hash: '#overview',
          title: 'Overview'
        },
        {
          hash: '#insights',
          title: 'Zone Insights'
        },
        // {
        //   hash: '#yaml',
        //   title: 'YAML'
        // }
      ],
      entity: [],
      rawEntity: null,
      yamlEntity: null,
      firstEntity: null,
      pageSize: this.$pageSize,
      pageOffset: null,
      next: null,
      hasNext: false,
      previous: [],
      tabGroupTitle: null,
      entityOverviewTitle: null,
      itemsPerCol: 3
    }
  },
  computed: {
    ...mapState({
      mesh: 'selectedMesh'
    }),
    ...mapGetters({
      multicluster: 'getMulticlusterStatus'
    }),
    // If you need to test multicluster without actually having it enabled
    // in Kuma, uncomment this and comment out the mapGetters above.
    // multicluster () {
    //   return true
    // },
    pageTitle () {
      const metaTitle = this.$route.meta.title

      return metaTitle
    },
    shareUrl () {
      const urlRoot = `${window.location.origin}#`
      const entity = this.entity

      const shareUrl = () => {
        if (this.$route.query.ns) {
          return this.$route.fullPath
        }

        return `${urlRoot}${this.$route.fullPath}?ns=${entity.name}`
      }

      return shareUrl()
    }
  },
  watch: {
    '$route' (to, from) {
      this.init()
    }
  },
  beforeMount () {
    this.init()
  },
  methods: {
    init () {
      if (this.multicluster) {
        this.loadData()
      }
    },
    goToPreviousPage () {
      this.pageOffset = this.previous.pop()
      this.next = null

      this.loadData()
    },
    goToNextPage () {
      this.previous.push(this.pageOffset)
      this.pageOffset = this.next
      this.next = null

      this.loadData()
    },
    tableAction (ev) {
      const data = ev

      // reset back to the first tab
      this.$store.dispatch('updateSelectedTab', this.tabs[0].hash)

      // set the active table row
      this.$store.dispatch('updateSelectedTableRow', data.name)

      // load the data into the tabs
      this.getEntity(data)
    },
    loadData () {
      this.isLoading = true
      this.isEmpty = false

      const endpoint = this.$api.getZoneStatus()

      const getZoneStatus = () => {
        return endpoint
          .then(response => {
            const nextCheck = (response && response.next) ? response.next : false

            // check to see if the `next` url is present
            if (nextCheck) {
              this.next = getOffset(response.next)
              this.hasNext = true
            } else {
              this.hasNext = false
            }

            const items = response

            if (items && items.length > 0) {
              // rewrite the status column to be more human-readable
              items.forEach(i => {
                const status = (i.active === false)
                  ? 'Offline'
                  : 'Online'

                delete i.active

                i.status = status
              })

              // sort the table data by name and the mesh it's associated with
              this.sortEntities(items)

              // set the first item as the default for initial load
              this.firstEntity = items[0].name

              // load the YAML entity for the first item on page load
              this.getEntity(items[0])

              // set the selected table row for the first item on page load
              this.$store.dispatch('updateSelectedTableRow', this.firstEntity)

              this.tableData.data = [...items]
              this.tableDataIsEmpty = false
              this.isEmpty = false
            } else {
              this.tableData.data = []
              this.tableDataIsEmpty = true
              this.isEmpty = true

              this.getEntity(null)
            }
          })
          .catch(error => {
            this.hasError = true
            this.isEmpty = true

            console.error(error)
          })
          .finally(() => {
            setTimeout(() => {
              this.isLoading = false
            }, process.env.VUE_APP_DATA_TIMEOUT)
          })
      }

      getZoneStatus()
    },
    getEntity (entity) {
      this.entityIsLoading = true
      this.entityIsEmpty = false

      const selected = ['type', 'name', 'mesh']

      const promise = new Promise((resolve, reject) => {
        if (entity && entity !== null) {
          resolve(entity)
        } else {
          const error = new Error('Entity is either undefined or not present.')

          reject(error)
        }
      })

      if (entity && entity !== null) {
        promise
          .then(response => {
            if (response) {
              // get the Zone details from the Zone Insights endpoint
              this.$api.getZoneOverview(response.name)
                .then((response) => {
                  this.tabGroupTitle = `Zone: ${response.name}`
                  this.entityOverviewTitle = `Zone Overview for ${response.name}`

                  // remove `zoneInsight` from the response since we already
                  // use it in the Zone Insights tab
                  const cleanRes = () => {
                    const src = Object.assign({}, response)

                    delete src.zoneInsight

                    return src
                  }

                  this.entity = getSome(response, selected)
                  this.rawEntity = stripTimes(response)
                  this.yamlEntity = cleanRes()
                })
            } else {
              this.entity = null
              this.entityIsEmpty = true
            }
          })
          .catch(error => {
            this.entityHasError = true
            console.error(error)
          })
          .finally(() => {
            setTimeout(() => {
              this.entityIsLoading = false
            }, process.env.VUE_APP_DATA_TIMEOUT)
          })
      } else {
        setTimeout(() => {
          this.entityIsEmpty = true
          this.entityIsLoading = false
        }, process.env.VUE_APP_DATA_TIMEOUT)
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.add-mesh-button {
  background-color: var(--logo-green) !important;
}
</style>
