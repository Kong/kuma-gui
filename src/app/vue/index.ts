import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
