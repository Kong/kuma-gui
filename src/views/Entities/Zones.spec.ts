import { screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'
import renderWithVuex from '@/testUtils/renderWithVuex'
import Vue from 'vue'
import Zones from './Zones.vue'

describe('Zones.vue', () => {
  it('renders snapshot when no multizone', () => {
    const { container } = renderWithVuex(Zones)

    expect(container).toMatchSnapshot()
  })

  it('renders snapshot when multizone', async () => {
    const { container } = renderWithVuex(Zones, {
      store: { modules: { config: { state: { clientConfig: { mode: 'global' } } } } },
    })

    await screen.findByText(/cluster-1/)
    await screen.findByText(/dpToken/)

    expect(container).toMatchSnapshot()
  })

  it('renders config of multizone', async () => {
    renderWithVuex(Zones, {
      store: { modules: { config: { state: { clientConfig: { mode: 'global' } } } } },
    })

    await screen.findByText(/cluster-1/)
    await screen.findByText(/dpToken/)

    await userEvent.click(screen.getByText('Config'))
    expect(await screen.findByText(/adminAccessLogPath/)).toBeInTheDocument()
  })
})
