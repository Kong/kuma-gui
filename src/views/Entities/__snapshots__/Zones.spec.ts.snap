// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zones.vue renders snapshot when multizone 1`] = `
<div>
  <div
    class="zones"
  >
    <div
      class="component-frame"
    >
      <div
        class="data-overview"
        data-testid="data-overview"
      >
        <div
          class="data-table-controls mb-2"
        >
           
          <button
            class="k-button ml-2 refresh-button small primary"
            data-v-90a15d32=""
            type="button"
          >
            <div
              class="refresh-icon"
              data-v-90a15d32=""
            >
              <svg
                data-v-90a15d32=""
                viewBox="0 0 36 36"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  data-v-90a15d32=""
                  fill="#fff"
                  fill-rule="nonzero"
                >
                  <path
                    d="M18 5.5a12.465 12.465 0 00-8.118 2.995 1.5 1.5 0 001.847 2.363l.115-.095A9.437 9.437 0 0118 8.5l.272.004a9.487 9.487 0 019.07 7.75l.04.246H25a.5.5 0 00-.416.777l4 6a.5.5 0 00.832 0l4-6 .04-.072A.5.5 0 0033 16.5h-2.601l-.017-.15C29.567 10.2 24.294 5.5 18 5.5zM2.584 18.723l-.04.072A.5.5 0 003 19.5h2.6l.018.15C6.433 25.8 11.706 30.5 18 30.5c3.013 0 5.873-1.076 8.118-2.995a1.5 1.5 0 00-1.847-2.363l-.115.095A9.437 9.437 0 0118 27.5l-.272-.004a9.487 9.487 0 01-9.07-7.75l-.041-.246H11a.5.5 0 00.416-.777l-4-6a.5.5 0 00-.832 0l-4 6z"
                    data-v-90a15d32=""
                  />
                </g>
              </svg>
            </div>
             
            <span
              data-v-90a15d32=""
            >
              Refresh
            </span>
          </button>
        </div>
         
        <div
          class="data-overview-content"
        >
          <!---->
           
          <div
            class="data-overview-table"
          >
            <table
              class="k-table micro-table has-hover is-clickable"
              data-v-59ad4099=""
            >
              <thead
                data-v-59ad4099=""
              >
                <tr
                  data-v-59ad4099=""
                >
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                    <!---->
                  </th>
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                     Status 
                  </th>
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                     Name 
                  </th>
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                     Zone CP Version 
                  </th>
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                     Backend 
                  </th>
                  <th
                    class=""
                    data-v-59ad4099=""
                  >
                    <!---->
                  </th>
                </tr>
              </thead>
              <tbody
                data-v-59ad4099=""
              >
                <tr
                  data-v-59ad4099=""
                >
                  <td
                    data-v-59ad4099=""
                  >
                    <a
                      class="data-table-action-link is-active"
                      data-v-59ad4099=""
                    >
                      <span
                        class="action-link__active-state"
                        data-v-59ad4099=""
                      >
                        
              âœ“
              
                        <span
                          class="sr-only"
                          data-v-59ad4099=""
                        >
                          
                Selected
              
                        </span>
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <div
                      class="entity-status"
                      data-v-59ad4099=""
                    >
                      <span
                        class="entity-status__dot"
                        data-v-59ad4099=""
                      />
                       
                      <span
                        class="entity-status__label"
                        data-v-59ad4099=""
                      >
                        Online
                      </span>
                    </div>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    cluster-1
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    1.0.0-rc2-211-g823fe8ce
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    memory
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <svg
                      class="kong-icon mr-1 kong-icon-warning"
                      data-v-59ad4099=""
                      data-v-77239f42=""
                      height="20"
                      role="img"
                      viewBox="0 0 18 16"
                      width="20"
                    >
                      <title
                        data-v-77239f42=""
                      >
                        warning
                      </title>
                      <g
                        data-v-77239f42=""
                      >
                        <path
                          d="M7.926.621a1.24 1.24 0 0 1 2.148 0l7.763 13.535A1.235 1.235 0 0 1 16.763 16H1.237a1.23 1.23 0 0 1-1.074-1.844L7.926.621z"
                          data-v-77239f42=""
                          fill="var(--yellow-400)"
                        />
                        <path
                          d="M10 11v2H8v-2h2zm0-7v5.5H8V4h2z"
                          data-v-77239f42=""
                          fill="#000"
                          fill-opacity=".9"
                          id="preserveColor"
                        />
                      </g>
                    </svg>
                  </td>
                </tr>
                <tr
                  data-v-59ad4099=""
                >
                  <td
                    data-v-59ad4099=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-59ad4099=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-59ad4099=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <div
                      class="entity-status"
                      data-v-59ad4099=""
                    >
                      <span
                        class="entity-status__dot"
                        data-v-59ad4099=""
                      />
                       
                      <span
                        class="entity-status__label"
                        data-v-59ad4099=""
                      >
                        Online
                      </span>
                    </div>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    zone-1
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    1.0.0-rc2-211-g823fe8ce
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    memory
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <svg
                      class="kong-icon mr-1 kong-icon-warning"
                      data-v-59ad4099=""
                      data-v-77239f42=""
                      height="20"
                      role="img"
                      viewBox="0 0 18 16"
                      width="20"
                    >
                      <title
                        data-v-77239f42=""
                      >
                        warning
                      </title>
                      <g
                        data-v-77239f42=""
                      >
                        <path
                          d="M7.926.621a1.24 1.24 0 0 1 2.148 0l7.763 13.535A1.235 1.235 0 0 1 16.763 16H1.237a1.23 1.23 0 0 1-1.074-1.844L7.926.621z"
                          data-v-77239f42=""
                          fill="var(--yellow-400)"
                        />
                        <path
                          d="M10 11v2H8v-2h2zm0-7v5.5H8V4h2z"
                          data-v-77239f42=""
                          fill="#000"
                          fill-opacity=".9"
                          id="preserveColor"
                        />
                      </g>
                    </svg>
                  </td>
                </tr>
                <tr
                  data-v-59ad4099=""
                >
                  <td
                    data-v-59ad4099=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-59ad4099=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-59ad4099=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <div
                      class="entity-status"
                      data-v-59ad4099=""
                    >
                      <span
                        class="entity-status__dot"
                        data-v-59ad4099=""
                      />
                       
                      <span
                        class="entity-status__label"
                        data-v-59ad4099=""
                      >
                        Online
                      </span>
                    </div>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    zone-2
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    1.0.0-rc2-211-g823fe8ce
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    memory
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <svg
                      class="kong-icon mr-1 kong-icon-warning"
                      data-v-59ad4099=""
                      data-v-77239f42=""
                      height="20"
                      role="img"
                      viewBox="0 0 18 16"
                      width="20"
                    >
                      <title
                        data-v-77239f42=""
                      >
                        warning
                      </title>
                      <g
                        data-v-77239f42=""
                      >
                        <path
                          d="M7.926.621a1.24 1.24 0 0 1 2.148 0l7.763 13.535A1.235 1.235 0 0 1 16.763 16H1.237a1.23 1.23 0 0 1-1.074-1.844L7.926.621z"
                          data-v-77239f42=""
                          fill="var(--yellow-400)"
                        />
                        <path
                          d="M10 11v2H8v-2h2zm0-7v5.5H8V4h2z"
                          data-v-77239f42=""
                          fill="#000"
                          fill-opacity=".9"
                          id="preserveColor"
                        />
                      </g>
                    </svg>
                  </td>
                </tr>
                <tr
                  data-v-59ad4099=""
                >
                  <td
                    data-v-59ad4099=""
                  >
                    <a
                      class="data-table-action-link"
                      data-v-59ad4099=""
                    >
                      <span
                        class="action-link__normal-state"
                        data-v-59ad4099=""
                      >
                        
              View
            
                      </span>
                    </a>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <div
                      class="entity-status"
                      data-v-59ad4099=""
                    >
                      <span
                        class="entity-status__dot"
                        data-v-59ad4099=""
                      />
                       
                      <span
                        class="entity-status__label"
                        data-v-59ad4099=""
                      >
                        Online
                      </span>
                    </div>
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    zone-3
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    1.0.0-rc2-211-g823fe8ce
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    memory
                  </td>
                  <td
                    data-v-59ad4099=""
                  >
                    <svg
                      class="kong-icon mr-1 kong-icon-warning"
                      data-v-59ad4099=""
                      data-v-77239f42=""
                      height="20"
                      role="img"
                      viewBox="0 0 18 16"
                      width="20"
                    >
                      <title
                        data-v-77239f42=""
                      >
                        warning
                      </title>
                      <g
                        data-v-77239f42=""
                      >
                        <path
                          d="M7.926.621a1.24 1.24 0 0 1 2.148 0l7.763 13.535A1.235 1.235 0 0 1 16.763 16H1.237a1.23 1.23 0 0 1-1.074-1.844L7.926.621z"
                          data-v-77239f42=""
                          fill="var(--yellow-400)"
                        />
                        <path
                          d="M10 11v2H8v-2h2zm0-7v5.5H8V4h2z"
                          data-v-77239f42=""
                          fill="#000"
                          fill-opacity=".9"
                          id="preserveColor"
                        />
                      </g>
                    </svg>
                  </td>
                </tr>
              </tbody>
            </table>
             
            <div
              class="pagination"
            >
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
           
          <!---->
        </div>
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="tab-container"
      >
        <header
          class="tab__header"
        >
          <div>
            <h3>
              Zone: cluster-1
            </h3>
          </div>
        </header>
         
        <div
          class="tab__content-container"
        >
          <div
            class="k-tabs"
            data-v-a38b664a=""
          >
            <ul
              aria-label="ktabs"
              data-v-a38b664a=""
              role="tablist"
            >
              <li
                aria-controls="panel-0"
                aria-selected="true"
                class="tab-item active"
                data-v-a38b664a=""
                id="overview-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Overview
                </a>
              </li>
              <li
                aria-controls="panel-1"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="insights-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Zone Insights
                </a>
              </li>
              <li
                aria-controls="panel-2"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="config-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  Config
                </a>
              </li>
              <li
                aria-controls="panel-3"
                aria-selected="false"
                class="tab-item"
                data-v-a38b664a=""
                id="warnings-tab"
                role="tab"
                tabindex="0"
              >
                <a
                  class="tab-link"
                  data-v-a38b664a=""
                >
                  <span
                    class="flex items-center with-warnings"
                    data-v-a38b664a=""
                  >
                    <svg
                      class="kong-icon mr-1 kong-icon-warning"
                      data-v-77239f42=""
                      data-v-a38b664a=""
                      height="16"
                      role="img"
                      viewBox="0 0 18 16"
                      width="16"
                    >
                      <title
                        data-v-77239f42=""
                      >
                        warning
                      </title>
                      <g
                        data-v-77239f42=""
                      >
                        <path
                          d="M7.926.621a1.24 1.24 0 0 1 2.148 0l7.763 13.535A1.235 1.235 0 0 1 16.763 16H1.237a1.23 1.23 0 0 1-1.074-1.844L7.926.621z"
                          data-v-77239f42=""
                          fill="var(--yellow-400)"
                        />
                        <path
                          d="M10 11v2H8v-2h2zm0-7v5.5H8V4h2z"
                          data-v-77239f42=""
                          fill="#000"
                          fill-opacity=".9"
                          id="preserveColor"
                        />
                      </g>
                    </svg>
                     
                    <span
                      data-v-a38b664a=""
                    >
                      
            Warnings
          
                    </span>
                  </span>
                </a>
              </li>
            </ul>
            <div
              aria-labelledby="overview-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-0"
              role="tabpanel"
              tabindex="0"
            >
              <div
                class="label-list"
                data-v-a38b664a=""
              >
                <div
                  class="label-list-content"
                >
                  <div
                    class="kong-card noBorder"
                    data-v-4ad076e4=""
                  >
                    <!---->
                    <div
                      class="k-card-body"
                      data-v-4ad076e4=""
                    >
                      <div
                        class="label-list__col-wrapper multi-col"
                        data-v-4ad076e4=""
                      >
                        <div
                          data-v-4ad076e4=""
                        >
                          <ul
                            data-v-4ad076e4=""
                          >
                            <li
                              data-v-4ad076e4=""
                            >
                              <h4
                                data-v-4ad076e4=""
                              >
                                
                  type
                
                              </h4>
                               
                              <p
                                data-v-4ad076e4=""
                              >
                                
                  ZoneOverview
                
                              </p>
                            </li>
                            <li
                              data-v-4ad076e4=""
                            >
                              <h4
                                data-v-4ad076e4=""
                              >
                                
                  name
                
                              </h4>
                               
                              <p
                                data-v-4ad076e4=""
                              >
                                
                  cluster-1
                
                              </p>
                            </li>
                            <li
                              data-v-4ad076e4=""
                            >
                              <!---->
                               
                              <p
                                data-v-4ad076e4=""
                              >
                                
                  
                
                              </p>
                            </li>
                            <li
                              data-v-4ad076e4=""
                            >
                              <h4
                                data-v-4ad076e4=""
                              >
                                
                  Authentication Type
                
                              </h4>
                               
                              <p
                                data-v-4ad076e4=""
                              >
                                
                  dpToken
                
                              </p>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                 
                <!---->
                 
                <!---->
                 
                <!---->
              </div>
            </div>
            <div
              aria-labelledby="insights-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-1"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
            <div
              aria-labelledby="config-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-2"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
            <div
              aria-labelledby="warnings-tab"
              class="tab-container"
              data-v-a38b664a=""
              id="panel-3"
              role="tabpanel"
              tabindex="0"
            >
              <!---->
            </div>
          </div>
           
          <div>
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Zones.vue renders snapshot when no multizone 1`] = `
<div>
  <div
    class="zones"
  >
    <div
      class="empty-state-wrapper global-api-status"
      data-v-7725326c=""
    >
      <div
        class="empty-state-title"
        data-v-7725326c=""
      >
        <!---->
        <h5
          data-v-7725326c=""
        >
          <svg
            class="kong-icon kong-icon--centered kong-icon-dangerCircle"
            data-v-77239f42=""
            data-v-7725326c=""
            height="64"
            role="img"
            viewBox="0 0 24 24"
            width="64"
          >
            <title
              data-v-77239f42=""
            >
              dangerCircle
            </title>
            <g
              data-v-77239f42=""
            >
              <path
                d="M0,8a8,8 0 1,0 16,0a8,8 0 1,0 -16,0"
                data-v-77239f42=""
                fill="#BFBFBF"
              />
              <path
                d="M9 11v2H7v-2h2zm0-8v6.5H7V3h2z"
                data-v-77239f42=""
                fill="#FFF"
                id="preserveColor"
              />
            </g>
          </svg>
          
      Kuma is running in Standalone mode.
    
        </h5>
      </div>
      <div
        class="empty-state-content"
        data-v-7725326c=""
      >
        <p
          data-v-7725326c=""
        >
          <p
            data-v-7725326c=""
          >
            
        To access this page, you must be running in 
            <strong
              data-v-7725326c=""
            >
              Multi-Zone
            </strong>
             mode.
      
          </p>
        </p>
        <p
          data-v-7725326c=""
        >
          <a
            class="k-button primary"
            data-v-7725326c=""
            data-v-90a15d32=""
            href="https://kuma.io/docs/0.6.0/documentation/deployments/"
            target="_blank"
            type="button"
          >
            
        Learn More
      
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
